core/validation/initial_check.py

```diff
@@ -0,0 +1,111 @@
+from typing import Any
+
+import numpy as np
+import pandas as pd
+
+import core.validation.failures as vf
+from core.const import EXPECTED_ROUND_THREE_SHEETS
+from core.validation.failures import NoInputFailure
+
+
+def extract_round_three_submission_details(
+    workbook: dict[str, pd.DataFrame]
+) -> dict[str, list[str]] | dict[str, dict[Any, Any]]:
+    """
+    Extract submission details for round three from the given workbook.
+
+    :param workbook: A dictionary where keys are sheet names and values are pandas DataFrames.
+    :return: A dictionary containing the extracted submission details.
+    """
+    details_dict = {"ValueChecks": {}, "NullChecks": {}}
+
+    missing_sheets = check_missing_sheets(EXPECTED_ROUND_THREE_SHEETS, workbook)
+
+    if missing_sheets:
+        return {"Missing Sheets": missing_sheets}
+
+    sheet_a1 = workbook.get("1 - Start Here")
+    if sheet_a1 is not None:
+        details_dict["ValueChecks"]["Form Version"] = (
+            sheet_a1.iloc[6][1],
+            {"Town Deals and Future High Streets Fund Reporting Template (v3.0)"},
+        )
+        details_dict["ValueChecks"]["Reporting Period"] = (sheet_a1.iloc[4][1], {"1 October 2022 to 31 March 2023"})
+
+    sheet_a2 = workbook.get("2 - Project Admin")
+    if sheet_a2 is not None:
+        details_dict["ValueChecks"]["Fund Type"] = (sheet_a2.iloc[5][4], {"Town_Deal", "Future_High_Street_Fund"})
+        details_dict["NullChecks"]["Place Name"] = sheet_a2.iloc[6][4]
```

Comment:
Just realised we'll get index errors if a spreadsheet that is wildly wrong tries to be ingest... e.g. df.iloc[3][1] will fall over if the df is tiny. Not sure if we want to account for this... The fact we pull named sheets out and don't run the index lookups if they don't exist will protect against this in the vast majority of cases.. 

One solution that wouldn't require much work could be to add a try except round the uses of `iloc` and if IndexError is found then just add the sheet as missing and continue.. 

What do you think? It seems a bit excessive but it is definitely a possible edge case

Reply:
The extraction function is written under the assumption that the form will be formatted in the same way for each round, which is why we may need bespoke extraction functions for different rounds. I haven't encountered any R3 forms that don't adhere to the specific format presumed by this extraction method. 

To hedge our bets we could do something like:
```
    sheet_a1 = workbook.get("1 - Start Here")
    sheet_a2 = workbook.get("2 - Project Admin")
    try:
        details_dict["ValueChecks"]["Form Version"] = (
            sheet_a1.iloc[6][1],
            {"Town Deals and Future High Streets Fund Reporting Template (v3.0)"},
        )
        details_dict["ValueChecks"]["Reporting Period"] = (sheet_a1.iloc[4][1], {"1 October 2022 to 31 March 2023"})
        details_dict["ValueChecks"]["Fund Type"] = (sheet_a2.iloc[5][4], {"Town_Deal", "Future_High_Street_Fund"})
        details_dict["NullChecks"]["Place Name"] = sheet_a2.iloc[6][4]
    except IndexError as e:
        return details_dict
```

As the missing sheet will already be caught above, and so we merely want to return the dictionary containing this. However, this won't account for the sheet being present but with insufficient data (e.g. if it just had headers), in which case it'd try and fail to locate the specified cells.

Another way would be as you suggested:

```
missing_sheets = []

    sheet_a1 = workbook.get("1 - Start Here")
    sheet_a2 = workbook.get("2 - Project Admin")
    try:
        details_dict["ValueChecks"]["Form Version"] = (
            sheet_a1.iloc[6][1],
            {"Town Deals and Future High Streets Fund Reporting Template (v3.0)"},
        )
        details_dict["ValueChecks"]["Reporting Period"] = (sheet_a1.iloc[4][1], {"1 October 2022 to 31 March 2023"})
    except IndexError:    
        missing_sheets.append("1 - Start Here")
    
    try:
        details_dict["ValueChecks"]["Fund Type"] = (sheet_a2.iloc[5][4], {"Town_Deal", "Future_High_Street_Fund"})
        details_dict["NullChecks"]["Place Name"] = sheet_a2.iloc[6][4]
    except IndexError:    
        missing_sheets.append("2 - Project Admin")
```
In this instance, we don't check to see if sheets are missing before trying to retrieve the values from the specified cells as otherwise we'd potentially be adding the same sheet twice to the list. We no longer iterate through all of the sheets to check if present, but that was only included after an edge case in QA testing, and as post-transformation validation captures the EmptySheet phenomenon it should be alright.

Realistically, this edge case will only happen if for whatever reason the user has sheets with some data in for sheets a1 and a2, but not enough, and as the sheet is protected and cannot be altered to delete most of the cells it's very unlikely to happen, but if you think of the above two solutions works well enough and helps protect against this I am happy to implement. What do you think is the best way forward?

Reply:
The first solution could still incorrectly pass validation. If neither sheet was empty, but didn't conform to the Round 3 structure and in doing so raised an Index Error, then the details dict could be returned with either no ValueChecks or some valid ValueChecks that would then still pass validation. 

The second ensures that we raise a failure. Although I'm not sure an EmptySheet really describes the problem. It's more like an invalid sheet.

Reply:
The second solution would fall over if there was a legitimate empty sheet now though, because we'd call iloc on None, which would raise a different exception to an Index Error. We'd also need to check if either sheet was None, and add a missing sheet in that case and skip the try except block in that case

Reply:
I've pushed a new solution, which separates missing sheets from invalid sheets. If there are any missing sheets, they will be returned before the code block that identifies values by cell position. This will prevent the programme calling iloc on None. Invalid sheets are then treated in the same way missing sheets are, only raising a different type of ValidationFailure.